backtracking/subarray_sum.cpp

```diff
@@ -0,0 +1,116 @@
+ * @details Subset sum(only continuous subsets) problem
+ * (https://en.wikipedia.org/wiki/Subset_sum_problem)
+ * @author [Swastika Gupta](https://github.com/Swastyy)
+ */
+
+#include <cassert>        /// for assert
+#include <iostream>       /// for IO operations
+#include <unordered_map>  /// for unordered_map
+#include <vector>         /// for std::vector
+
+/**
+ * @namespace backtracking
+ * @brief subarray sum algorithm
+ */
+namespace backtracking {
+/**
+ * @namespace subarraySum
+ * @brief Functions for the [Subset
+ * sum](https://en.wikipedia.org/wiki/Subset_sum_problem) implementation
+ */
+namespace subarraySum {
+/**
+ * @brief The main function that implements the count of the subarrays
+ * @param sum is the required sum of any subarrays
+ * @param in_arr is the input array
+ * @returns count of the number of subsets with required sum
+ */
+
+uint64_t subarray_sum(int sum, const std::vector<int> &in_arr) {
+    int nelement = in_arr.size();
+    int count_of_subset = 0;
+    int current_sum = 0;
+    std::unordered_map<int, int> sumarray;  // to store the subarrays count
+                                            // frequency having some sum value
+
+    for (int i = 0; i < nelement; i++) {
+        current_sum += in_arr[i];
+
+        if (current_sum == sum) {
+            count_of_subset++;
+        }
+        // If in case current_sum is greater than the required sum
+        if (sumarray.find(current_sum - sum) != sumarray.end()) {
+            count_of_subset += (sumarray[current_sum - sum]);
+        }
+        sumarray[current_sum]++;
+    }
+    return count_of_subset;
+}
+}  // namespace subarraySum
+}  // namespace backtracking
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    // Test 1
+    std::cout << "1st test ";
+    std::vector<int> array1 = {-7, -3, -2, 5, 8};  // input array
+    assert(
+        backtracking::subarraySum::subarray_sum(0, array1) ==
+        1);  // first argument in subarray_sum function is the required sum and
+             // second is the input array, answer is the subarray {(-3,-2,5)}
+    std::cout << "passed" << std::endl;
+
+    // Test 2
+    std::cout << "2nd test ";
+    std::vector<int> array2 = {1, 2, 3, 3};
+    assert(backtracking::subarraySum::subarray_sum(6, array2) ==
+           2);  // here we are expecting 2 subsets which sum up to 6 i.e.
+                // {(1,2,3),(3,3)}
+    std::cout << "passed" << std::endl;
+
+    // Test 3
+    std::cout << "3rd test ";
+    std::vector<int> array3 = {1, 1, 1, 1};
+    assert(backtracking::subarraySum::subarray_sum(1, array3) ==
+           4);  // here we are expecting 4 subsets which sum up to 1 i.e.
+                // {(1),(1),(1),(1)}
+    std::cout << "passed" << std::endl;
+
+    // Test 4
+    std::cout << "4th test ";
+    std::vector<int> array4 = {3, 3, 3, 3};
+    assert(backtracking::subarraySum::subarray_sum(6, array4) ==
```

Comment:
```suggestion
    assert(backtracking::subarray_sum::subarray_sum(6, array4) ==
```

Reply:
Is it okay to have the namespace and function name same?

Reply:
> Is it okay to have the namespace and function name same?

Yes, there's no problem with that. 🙂